### ISSUE TITLE
Possible to set collection parents before adding items to collection in Many relation?

### ISSUE THREAD

**Issue by jpslav** [dhruvaray/backbone-associations#93]
Howdy -

In the source, when setting up a Many relation, it looks like the `parents` property of a collection in a `Many` relation is set after the contents of the collection are set (roughly lines 286 and 234, respectively).

Would it be feasible to assign the parents first and set the contents second?

I have a Model `A` that has many `B`s and `C`s.  Each `C` can reference many `B`s (something of a composite pattern going on here).  When editing an `A`, the `B`s and `C`s are on the screen at the same time.  In order to properly display each `C` I need that `C` to be aware of the `A`'s collection of `B`s (so that that `C` can: know which `B`s it can point to, listen for deletion of `B`s, etc).  

Easy as ABC, right? ;-)

In a number of situations, there is no problem here.  I can take a `C` and find the `A` that contains it using `myC.collection.parents[0]`.  With that `A` I can then get the collection of `B`s.  

However, I have a problem in the comparator in the `C` collection, in that it needs to know some of the properties of the elements in the `B` collection.  But the comparator is being called when elements are added to the collection, which (per the above) appears to happen before the `C` collection parent is set, which means the `C`s can't get at the relevant `B`s.

Is moving the setting of the parents up earlier feasible?

Still loving this library.

**Comment by dhruvaray** [comment]
Thanks!

It may be difficult to do in the form you ask... Would this work for you?

Pass in options - `{sort:false}` when you call the set/add method. On completion of the set/add operation, you can do C.sort() (in your code) which would always work as the graph would be completely set up.

**Comment by jpslav** [comment]
I don't think I can pass in `{sort: false}` because the setting/adding of the `B`s and `C`s is happening during the initial loading of a larger nested JSON document, e.g.

```
K
  L
    an A
      collection of Bs
      collection of Cs
    another A
      ...
```

So I'm creating an `L` model passing in this bigger document, meaning the sets/adds are happening inside backbone-associations. 

I have for the moment done something similar to what you suggested though.  My comparator essentially bails out if the parent information isn't set yet.  Then in my view code, when I've navigated down to the part that displays the C collection, I sort it before adding it to the view.  This works, but maybe isn't ideal as I have to remember to stick a `sort` in at the appropriate place in some view code.

If setting the parent prior to adding/setting the collection items isn't doable, would it be possible to fire an event that indicates that the parent has been set?  That way I can have the collection listen for it and then re-sort itself.  As it is now all of the BB events (add, reset, etc) happen before the parent is set so there's nothing for me to react to after the parent is set.

**Comment by dhruvaray** [comment]
I agree... I have just made a commit which set's the parents earlier. Can you try it an let me know?

**Comment by jpslav** [comment]
Thanks for the commit.  I tried it and it does indeed set the collection's parent before adding models to it, _but_ when I ran it I realized that this fix doesn't really get me where I am trying to go.   

The `set` / `_set` / `_setAttr` code looks to be setting up the models with a depth-first approach (which makes sense) -- a model's relations are built up and then set on that model in the last line of `_setAttr`.  This means that while this commit does give me a properly-set `parent` before items are added to relations, that `parent` doesn't yet have its associations set (it won't be until that last line of `_setAttr`).  And it is these associations that I'm after (I want to go from a child up to its parent and back down to another child).

I don't see a trivial way around this.  The only idea that pops to mind would be to use a two pass approach: one that sets up the relations (with parents pointing to children and vice versa) and then another that goes and sets the content on those relations.  But of course, this sounds like a big change :-)

**Comment by jpslav** [comment]
I was also just looking at how the code sometimes sets `_deferEvents` to true.  Thought maybe that adding that right after `data = collectionType ? new collectionType() : this._createCollection(relatedModel);` could help me out (letting me attach to events after everything loaded) but I haven't had luck with this.

**Comment by dhruvaray** [comment]
> > I was also just looking at how the code sometimes sets _deferEvents to true. 

The _deferEvents is only to make sure that events are not fired until the entire update is done

> > The set / _set / _setAttr code looks to be setting up the models with a depth-first approach (which makes sense) -

Yes. So it is not really that useful. So I am going to revert back that change.

> > The only idea that pops to mind would be to use a two pass approach: one that sets 

Sorry...Don't see that happening!

Why don't you build up your graph in stages (via 2/3 set operations). First `set` the A's on L. Then the B's and C's can be `set` simultaneously on A (assuming B & Cs are siblings of A)

**Comment by jpslav** [comment]
Thanks, @dhruvaray.  Totally agree that a 2-pass solution (or something similar) is not in the cards.

Re the suggestion to build up the graph in stages, my preference is just to create the top-level parent giving it the entire object graph and have it (and its children and so on) build up the required instances all at once.  And thanks to you for making this library so I can do that :-)

FWIW, I did find another solution that I think is pretty decent.  In all of this, I'm trying to keep as much logic in the model code as possible (to avoid having to put a random `sort()` call in the right place in some view code, as mentioned above).  In the `A` model, I had been trying to `listenTo` events for when its `B`s and `C`s changed.  I originally was trying to setup these listeners in `A`'s `initialize` method -- but I was never seeing the events like I thought I would.  Then I realized that that `initialize` method is being called after the `B` and `C` collections are being set (the `initialize` call is the last step in the `Backbone.Model` constructor).  So what I've done is just to override `A`'s constructor to add these listeners early on in the process:

```
class A extends Backbone.AssociatedModel
  relations: [
    {
      type:  Backbone.Many,
      key: 'as',
      collectionType: 'As'
    },
    {
      type: Backbone.Many,
      key: 'bs',
      collectionType: 'Bs'
    }
  ]

  constructor: () ->
    @on 'change:as', () -> @get('bs').sort()
    super
```

The backbone docs talk about overriding the constructor only "if you're looking to get fancy", but there is at least one other discussion out there saying this is an ok approach: https://github.com/jashkenas/backbone/issues/2332#issuecomment-14405988
